Mark book as read

ramyagsri opened the pull request:
Added functionality to allow user to mark a book from reading list as finished
1) Added new PUT /api/reading-list/:id/finished endpoint in reading-list controller
2) Added new actions markBookAsFinished, markBookAsFinishedSuccess, markBookAsFinishedFailure corresponding to new PUT api request
3) Created effect and reducer function for markBookAsFinished action
4) Made changes to UI such that user has option to mark a book as read, and once he marks it finished message with date will be displayed on page
5) Added e2e spec to cover mark as finished feature

Demo of functionality: https://recordit.co/gygiPOhtLa
e2e test: https://recordit.co/dIECJzHQuH
<img width="1616" alt="Screen Shot 2020-09-30 at 1 11 33 PM" src="https://user-images.githubusercontent.com/19898265/94723528-7f070200-031e-11eb-8463-e61822ed0efe.png">

ramyagsri commented:
@jdpearce This task is ready for review. Please check

catfireparty commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -17,4 +17,31 @@ describe('When: I use the reading list feature', () => {
+    await browser.get('/');
+    await browser.wait(
+      ExpectedConditions.textToBePresentInElement($('tmo-root'), 'okreads')
+    );
```
Again, why not put these in a beforeEach?

ramyagsri replied:
Done, moved it to `beforeEach`.

catfireparty commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -17,4 +17,31 @@ describe('When: I use the reading list feature', () => {
+    await form.submit();
+
+    // add book to reading list
+    await $('[data-testing="add-to-reading-list"]').click();
```
As in Task 3, this throws a warning - better to use the element array finder and select the first enabled button.

ramyagsri replied:
Updated same as in task 3

catfireparty commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -17,4 +17,31 @@ describe('When: I use the reading list feature', () => {
+    await readingListToggle.click();
+
+    // mark as read
+    const markAsFinishedButton = await element(by.css('.mark-finish-text'));
```
This also throws a warning...

catfireparty commented on libs/api/books/src/lib/reading-list.service.ts:
```diff
@@ -28,4 +28,14 @@ export class ReadingListService {
+  async markBookAsFinished(id: string, item: ReadingListItem): Promise<void> {
+    this.storage.update(list => {
+      return list.filter(listItem => listItem.bookId !== id);
+      const itemToBeMarked = list.find( listItem => listItem.bookId === id);
```
You've got unreachable code here...this means that, while the front end may look ok because of the optimistic updating, the backend never gets the update.

ramyagsri replied:
Made updated to code to make sure no unreachable code is present.

catfireparty commented on libs/books/data-access/src/lib/+state/reading-list.reducer.ts:
```diff
@@ -58,7 +58,17 @@ const readingListReducer = createReducer(
+      id: action.item.bookId,
+      changes: {
+        finished: true,
+        finishedDate: new Date().toISOString()
```
Are you sure you should be setting this in a reducer? Ideally this should already be in the action when you receive it. Setting it here and in the effect is going to result in two values that are out of sync.

ramyagsri replied:
Addressed it

catfireparty commented on libs/books/feature/src/lib/book-search/book-search.component.html:
```diff
@@ -32,12 +32,13 @@
               color="primary"
               (click)="addBookToReadingList(book)"
-              [disabled]="book.isAdded"
+              [disabled]="book.isAdded || book.finished"
```
Do you need to do this? Surely a book that is finished is _also_ added?

ramyagsri replied:
Right. Reverted to book.isAdded check.

catfireparty commented on libs/books/feature/src/lib/reading-list/reading-list.component.html:
```diff
@@ -19,6 +19,18 @@
       </button>
     </div>
+    <div class="finished-text" *ngIf="readingItem.finished">
+      Finished on: {{ readingItem.finishedDate | date:'MM/dd/yyyy' }}
```
I like that you've formatted the date but remember that only the US uses this format. Ideally you would either figure out the users's region or just use ISO8601

ramyagsri replied:
Addressed.
<img width="1613" alt="Screen Shot 2020-10-01 at 9 37 12 PM" src="https://user-images.githubusercontent.com/19898265/94883204-a8fc1980-042f-11eb-84f4-0c3c0b7cb17f.png">

catfireparty commented on libs/books/feature/src/lib/reading-list/reading-list.component.html:
```diff
@@ -19,6 +19,18 @@
+      *ngIf="!readingItem.finished"
+      (click)="markAsFinished(readingItem)"
+    >
+      Mark as read <mat-icon class="mark-finish-circle">check_circle</mat-icon>
```
You've got a label here _and_ the aria label. Do you need both?

ramyagsri replied:
Removed the aria label

catfireparty commented on libs/books/feature/src/lib/reading-list/reading-list.component.ts:
```diff
@@ -15,4 +15,8 @@ export class ReadingListComponent {
   }
+
+  public markAsFinished(item): void {
+    this.store.dispatch(markBookAsFinished({ item, finishedDate:  new Date().toISOString() }));
```
So you are setting this value in the action after all! Why set it in all the other places too?

ramyagsri replied:
I made changes at other places to use it from action

catfireparty reviewed (changes requested):
Nearly there. I think there's a few things to tidy up though.

ramyagsri commented:
@jdpearce Made changes, please review again

mandarini commented on apps/okreads-e2e/src/specs/reading-list.spec.ts:
```diff
@@ -17,4 +19,33 @@ describe('When: I use the reading list feature', () => {
     );
   });
+
+  it('Then: I should see mark book as finished', async () => {
```
Your test fails:
```
1) When: I use the reading list feature Then: I should see mark book as finished
  Message:
    Failed: element not interactable
```

(it cannot find the element you want it to click)

ramyagsri replied:
My bad, missed to include open reading list before marking book as read. Updated e2e spec. Demo of e2e test - https://recordit.co/PPy9C2GTIc

mandarini commented on libs/books/data-access/src/lib/+state/reading-list.actions.ts:
```diff
@@ -41,3 +41,18 @@ export const confirmedRemoveFromReadingList = createAction(
+  props<{ item: ReadingListItem; finishedDate: string }>()
+);
+
+export const markBookAsFinishedSuccess = createAction(
```
Not that important, but I would keep the naming convention of the rest of the actions (`confirmedMarkBookAsFinished` as `confirmedAddToReadingList`

ramyagsri replied:
Addressed.

mandarini commented on libs/books/data-access/src/lib/+state/reading-list.actions.ts:
```diff
@@ -41,3 +41,18 @@ export const confirmedRemoveFromReadingList = createAction(
+  props<{ item: ReadingListItem }>()
+);
+
+export const markBookAsFinishedFailure = createAction(
```
Same as above, I would keep the naming convention about the failed actions.

ramyagsri replied:
Updated accordingly

mandarini commented on libs/books/data-access/src/lib/+state/reading-list.actions.ts:
```diff
@@ -41,3 +41,18 @@ export const confirmedRemoveFromReadingList = createAction(
+);
+
+export const markBookAsFinishedSuccess = createAction(
+  '[Reading List API] Mark book as finished success',
```
Same here about the conventions, see line 26 of same file. Again, not that important, but it's nice, shows awareness of conventions etc

ramyagsri replied:
Updated accordingly

mandarini commented on libs/books/feature/src/lib/reading-list/reading-list.component.html:
```diff
@@ -1,24 +1,36 @@
+  <section *ngFor="let readingItem of readingList$ | async">
+    <div class="reading-list-item">
+      <div>
+        <img class="reading-list-item--cover" [src]="readingItem.coverUrl" alt="" />
```
Should this `alt` attribute be there if it's empty? And since it's there, should it be empty?

ramyagsri replied:
This was changed as part of first task, Added empty alt attribute for decorative image (as the text shown in image is already presented to its right) - Referred from https://www.w3.org/WAI/tutorials/images/decision-tree/

mandarini commented on libs/books/feature/src/lib/reading-list/reading-list.component.ts:
```diff
@@ -15,4 +15,8 @@ export class ReadingListComponent {
   }
+
+  markAsFinished(item): void {
+    this.store.dispatch(markBookAsFinished({ item, finishedDate:  new Date().toLocaleString() }));
```
Should you be using another native Date method here, to make sure you're getting the  ISO date string?

ramyagsri replied:
Updated accordingly.

<img width="1615" alt="Screen Shot 2020-10-05 at 7 06 52 PM" src="https://user-images.githubusercontent.com/19898265/95144298-f03f1e80-073d-11eb-8791-a92d9c1bb0b8.png">

mandarini reviewed (commented):
Test fails.. Also, check Date for ISO date string.

ramyagsri committed:
mark as read

ramyagsri commented:
@mandarini Please review the updated changes. 
Changes made as part of last commit 
1. Renamed some of the actions to match with other naming convention
2. Updated Date method to fetch ISO date 
3. Updated e2e specs
